// angular-tree-grid/esm2020/lib/modules/tree-cell/tree-cell.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TreeCellComponent } from './tree-cell.component';
import { TreeCellViewComponent } from './components/tree-cell-view/default/tree-cell-view.component';
import { CustomCellViewComponent } from './components/tree-cell-view/custom/custom-tree-cell.component';
import { TreeCellActionsComponent } from './components/tree-cell-actions/tree-cell-actions.component';
import { CustomCellEditorComponent } from './components/tree-cell-editor/custom/custom-tree-cell-Editor.component';
import { TreeCellEditorComponent } from './components/tree-cell-editor/default/cell-editor/tree-cell-editor.component';
import { FormsModule } from '@angular/forms';
import { SafeHtmlPipe } from '../../pipes/pipe.safehtml';
import * as i0 from "@angular/core";
export class TreeCellModule {
}
TreeCellModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: TreeCellModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TreeCellModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: TreeCellModule, declarations: [TreeCellComponent,
        TreeCellViewComponent,
        CustomCellViewComponent,
        TreeCellActionsComponent,
        CustomCellEditorComponent,
        TreeCellEditorComponent,
        SafeHtmlPipe], imports: [CommonModule,
        FormsModule], exports: [TreeCellComponent,
        CustomCellViewComponent,
        CustomCellEditorComponent,
        TreeCellEditorComponent,
        TreeCellActionsComponent,
        SafeHtmlPipe] });
TreeCellModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: TreeCellModule, imports: [[
            CommonModule,
            FormsModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: TreeCellModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        TreeCellComponent,
                        TreeCellViewComponent,
                        CustomCellViewComponent,
                        TreeCellActionsComponent,
                        CustomCellEditorComponent,
                        TreeCellEditorComponent,
                        SafeHtmlPipe
                    ],
                    imports: [
                        CommonModule,
                        FormsModule
                    ],
                    exports: [
                        TreeCellComponent,
                        CustomCellViewComponent,
                        CustomCellEditorComponent,
                        TreeCellEditorComponent,
                        TreeCellActionsComponent,
                        SafeHtmlPipe
                    ]
                }]
        }] });
