// angular-tree-grid/esm2020/lib/modules/tree-head/tree-head.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TreeHeadComponent } from './tree-head.component';
import * as i0 from "@angular/core";
export class TreeHeadModule {
}
TreeHeadModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: TreeHeadModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TreeHeadModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: TreeHeadModule, declarations: [TreeHeadComponent], imports: [CommonModule], exports: [TreeHeadComponent] });
TreeHeadModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: TreeHeadModule, imports: [[
            CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: TreeHeadModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TreeHeadComponent],
                    imports: [
                        CommonModule
                    ],
                    exports: [TreeHeadComponent]
                }]
        }] });
