// angular-tree-grid/esm2020/lib/modules/tree-body/tree-body.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TreeBodyComponent } from './tree-body.component';
import { TreeCellModule } from '../tree-cell/tree-cell.module';
import { FormsModule } from '@angular/forms';
import { AddRowComponent } from './components/add-row/add-row.component';
import { FilterRowComponent } from './components/filter-row/filter-row.component';
import { SubgridModule } from '../subgrid/subgrid.module';
import * as i0 from "@angular/core";
export class TreeBodyModule {
}
TreeBodyModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: TreeBodyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TreeBodyModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: TreeBodyModule, declarations: [TreeBodyComponent, AddRowComponent, FilterRowComponent], imports: [CommonModule,
        TreeCellModule,
        FormsModule,
        SubgridModule], exports: [TreeBodyComponent] });
TreeBodyModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: TreeBodyModule, imports: [[
            CommonModule,
            TreeCellModule,
            FormsModule,
            SubgridModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: TreeBodyModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TreeBodyComponent, AddRowComponent, FilterRowComponent],
                    imports: [
                        CommonModule,
                        TreeCellModule,
                        FormsModule,
                        SubgridModule
                    ],
                    exports: [TreeBodyComponent]
                }]
        }] });
