// angular-tree-grid/esm2020/lib/angular-tree-grid.module.mjs
import { NgModule } from '@angular/core';
import { AngularTreeGridComponent } from './angular-tree-grid.component';
import { CommonModule } from '@angular/common';
import { TreeBodyModule } from './modules/tree-body/tree-body.module';
import { TreeHeadModule } from './modules/tree-head/tree-head.module';
import * as i0 from "@angular/core";
export class AngularTreeGridModule {
}
AngularTreeGridModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: AngularTreeGridModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AngularTreeGridModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: AngularTreeGridModule, declarations: [AngularTreeGridComponent], imports: [CommonModule,
        TreeBodyModule,
        TreeHeadModule], exports: [AngularTreeGridComponent] });
AngularTreeGridModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: AngularTreeGridModule, imports: [[
            CommonModule,
            TreeBodyModule,
            TreeHeadModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: AngularTreeGridModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [AngularTreeGridComponent],
                    imports: [
                        CommonModule,
                        TreeBodyModule,
                        TreeHeadModule
                    ],
                    exports: [AngularTreeGridComponent]
                }]
        }] });
export { DefaultEditor } from './modules/tree-cell/components/tree-cell-editor/default/default-editor.component';
