// angular-tree-grid/esm2020/lib/modules/tree-cell/components/tree-cell-view/custom/custom-tree-cell.component.mjs
import { Component, Input, ViewChild, ViewContainerRef, } from '@angular/core';
import * as i0 from "@angular/core";
export class CustomCellViewComponent {
    constructor(resolver) {
        this.resolver = resolver;
    }
    ngOnInit() {
        if (this.column.component && !this.custom_component) {
            this.createCustomComponent();
            this.callOnComponentInit();
        }
    }
    ngOnDestroy() {
        if (this.custom_component) {
            this.custom_component.destroy();
        }
    }
    createCustomComponent() {
        const componentFactory = this.resolver.resolveComponentFactory(this.column.component);
        this.custom_component = this.custom_view.createComponent(componentFactory);
    }
    callOnComponentInit() {
        this.column.on_component_init && this.column.on_component_init(this.custom_component.instance);
        this.custom_component.instance.cell_value = this.row_data[this.column.name];
        this.custom_component.instance.row_data = this.row_data;
        this.custom_component.instance.column = this.column;
    }
}
CustomCellViewComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: CustomCellViewComponent, deps: [{ token: i0.ComponentFactoryResolver }], target: i0.ɵɵFactoryTarget.Component });
CustomCellViewComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.2.1", type: CustomCellViewComponent, selector: "db-custom-cell-component", inputs: { column: "column", row_data: "row_data" }, viewQueries: [{ propertyName: "custom_view", first: true, predicate: ["customView"], descendants: true, read: ViewContainerRef, static: true }], ngImport: i0, template: `
      <ng-template #customView></ng-template>
    `, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: CustomCellViewComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'db-custom-cell-component',
                    template: `
      <ng-template #customView></ng-template>
    `,
                }]
        }], ctorParameters: function () { return [{ type: i0.ComponentFactoryResolver }]; }, propDecorators: { column: [{
                type: Input
            }], row_data: [{
                type: Input
            }], custom_view: [{
                type: ViewChild,
                args: ['customView', { read: ViewContainerRef, static: true }]
            }] } });
