// angular-tree-grid/esm2020/lib/modules/subgrid/subgrid.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SubgridComponent } from './subgrid.component';
import { TreeCellModule } from '../tree-cell/tree-cell.module';
import { SubgridHeadComponent } from './components/subgrid-head/subgrid-head.component';
import { SubgridBodyComponent } from './components/subgrid-body/subgrid-body.component';
import * as i0 from "@angular/core";
export class SubgridModule {
}
SubgridModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: SubgridModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SubgridModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: SubgridModule, declarations: [SubgridComponent, SubgridHeadComponent, SubgridBodyComponent], imports: [CommonModule,
        TreeCellModule], exports: [SubgridComponent] });
SubgridModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: SubgridModule, imports: [[
            CommonModule,
            TreeCellModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: SubgridModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [SubgridComponent, SubgridHeadComponent, SubgridBodyComponent],
                    imports: [
                        CommonModule,
                        TreeCellModule
                    ],
                    exports: [SubgridComponent]
                }]
        }] });
