// angular-tree-grid/esm2020/lib/modules/tree-cell/components/tree-cell-editor/custom/custom-tree-cell-Editor.component.mjs
import { Component, Input, ViewChild, ViewContainerRef, } from '@angular/core';
import { DefaultEditor } from '../default/default-editor.component';
import * as i0 from "@angular/core";
export class CustomCellEditorComponent extends DefaultEditor {
    constructor(resolver) {
        super();
        this.resolver = resolver;
    }
    ngOnInit() {
        if (this.column.editor && !this.custom_component) {
            this.createCustomComponent();
            this.callOnComponentInit();
        }
    }
    ngOnDestroy() {
        if (this.custom_component) {
            this.custom_component.destroy();
        }
    }
    createCustomComponent() {
        const componentFactory = this.resolver.resolveComponentFactory(this.column.editor);
        this.custom_component = this.custom_view.createComponent(componentFactory);
    }
    callOnComponentInit() {
        this.column.on_component_init && this.column.on_component_init(this.custom_component.instance);
        this.custom_component.instance.cell_value = this.cell_value;
        this.custom_component.instance.row_data = this.row_data;
        this.custom_component.instance.column = this.column;
        this.custom_component.instance.editcomplete.subscribe((event) => this.editcomplete.emit(event));
        this.custom_component.instance.canceledit.subscribe((event) => this.canceledit.emit(event));
        this.custom_component.instance.cellclick.subscribe((event) => this.cellclick.emit(event));
    }
}
CustomCellEditorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: CustomCellEditorComponent, deps: [{ token: i0.ComponentFactoryResolver }], target: i0.ɵɵFactoryTarget.Component });
CustomCellEditorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.2.1", type: CustomCellEditorComponent, selector: "db-custom-cell-editor-component", inputs: { column: "column", cell_value: "cell_value", row_data: "row_data" }, viewQueries: [{ propertyName: "custom_view", first: true, predicate: ["customView"], descendants: true, read: ViewContainerRef, static: true }], usesInheritance: true, ngImport: i0, template: `
      <ng-template #customView></ng-template>
    `, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: CustomCellEditorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'db-custom-cell-editor-component',
                    template: `
      <ng-template #customView></ng-template>
    `,
                }]
        }], ctorParameters: function () { return [{ type: i0.ComponentFactoryResolver }]; }, propDecorators: { column: [{
                type: Input
            }], cell_value: [{
                type: Input
            }], row_data: [{
                type: Input
            }], custom_view: [{
                type: ViewChild,
                args: ['customView', { read: ViewContainerRef, static: true }]
            }] } });
