// angular-tree-grid/esm2020/lib/modules/tree-cell/components/tree-cell-editor/default/default-editor.component.mjs
import { Output, EventEmitter, Component } from '@angular/core';
import * as i0 from "@angular/core";
export class DefaultEditor {
    constructor() {
        this.canceledit = new EventEmitter();
        this.editcomplete = new EventEmitter();
        this.cellclick = new EventEmitter();
    }
}
DefaultEditor.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: DefaultEditor, deps: [], target: i0.ɵɵFactoryTarget.Component });
DefaultEditor.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.2.1", type: DefaultEditor, selector: "ng-component", outputs: { canceledit: "canceledit", editcomplete: "editcomplete", cellclick: "cellclick" }, ngImport: i0, template: '', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.1", ngImport: i0, type: DefaultEditor, decorators: [{
            type: Component,
            args: [{
                    selector: '',
                    template: '',
                }]
        }], propDecorators: { canceledit: [{
                type: Output
            }], editcomplete: [{
                type: Output
            }], cellclick: [{
                type: Output
            }] } });
